server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java

```diff
@@ -0,0 +1,235 @@
+            super(settings, nodeEnvironment, new MetaStateService(nodeEnvironment, xContentRegistry()),
+                    Mockito.mock(MetaDataIndexUpgradeService.class), Mockito.mock(MetaDataUpgrader.class),
+                    transportService);
+        }
+
+        @Override
+        protected void upgradeMetaData(MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader) {
+            // MetaData upgrade is tested in GatewayMetaStateTests, we override this method to NOP to make mocking easier
+        }
+    }
+
+    private NodeEnvironment nodeEnvironment;
+    private ClusterName clusterName;
+    private Settings settings;
+    private TransportService transportService;
+    private DiscoveryNode localNode;
+
+    @Override
+    public void setUp() throws Exception {
+        nodeEnvironment = newNodeEnvironment();
+        localNode = new DiscoveryNode("node1", buildNewFakeTransportAddress(), Collections.emptyMap(),
+                Sets.newHashSet(DiscoveryNode.Role.MASTER), Version.CURRENT);
+        clusterName = new ClusterName(randomAlphaOfLength(10));
+        settings = Settings.builder().put(ClusterName.CLUSTER_NAME_SETTING.getKey(), clusterName.value()).build();
+        transportService = Mockito.mock(TransportService.class);
+        Mockito.when(transportService.getLocalNode()).thenReturn(localNode);
+        super.setUp();
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        nodeEnvironment.close();
+        super.tearDown();
+    }
+
+    private Gateway newGateway() throws IOException {
+        return new Gateway(settings, nodeEnvironment, transportService);
+    }
+
+    private Gateway maybeNew(Gateway gateway) throws IOException {
+        if (randomBoolean()) {
+            return newGateway();
+        }
+        return gateway;
+    }
+
+    public void testInitialState() throws IOException {
+        Gateway gateway = newGateway();
+        ClusterState state = gateway.getLastAcceptedState();
+        assertThat(state.getClusterName(), equalTo(clusterName));
+        assertTrue(MetaData.isGlobalStateEquals(state.metaData(), MetaData.EMPTY_META_DATA));
+        assertThat(state.getVersion(), equalTo(Manifest.empty().getClusterStateVersion()));
+        assertThat(state.getNodes().getLocalNode(), equalTo(localNode));
+
+        long currentTerm = gateway.getCurrentTerm();
+        assertThat(currentTerm, equalTo(Manifest.empty().getCurrentTerm()));
+    }
+
+    public void testSetCurrentTerm() throws IOException {
+        Gateway gateway = newGateway();
+
+        for (int i = 0; i < randomIntBetween(1, 5); i++) {
+            final long currentTerm = randomNonNegativeLong();
+            gateway.setCurrentTerm(currentTerm);
+            gateway = maybeNew(gateway);
+            assertThat(gateway.getCurrentTerm(), equalTo(currentTerm));
```

Comment:
also check that the cluster state remains untouched?

Reply:
See below